Fix destruction of core nodejs modules

eyalroth opened the pull request:
## Summary

Fixes #15638 that was introduced in #15215.

The PR fixes the deletion of core modules that are set as globals (some third party libraries do that for some reason) by adding protection on them from deletion when they are required.

## Test plan

The `jest-environment-node` unit tests were expanded to catch the bug.

Plus I added a new yarn script `test-node-env` (+ added to CI) to make sure that the `jest-environment-node` unit tests run in the same CI shard; if they don't run in the same process/jest execution, then they don't properly test the bug (same for #15636).

cpojer commented:
@eyalroth really appreciate all your work here. I'm worried that the original change is too invasive, and that it would be problematic in more states and/or hold back the Jest 30 release too much. At the same time, it is indeed a critical memory fix. Please let me know if you'll be able to resolve this soon so we can move forward with Jest 30.

eyalroth commented:
@cpojer I totally understand.

I will be updating the PR later today with a more robust fix (applying the protection code recursively).

However, I'm also a bit wary that this won't catch all potential bugs with the change.

How about adding a flag to control this feature (either opt in or out)?

eyalroth commented:
@cpojer the aim is only to protect the core modules recursively, and make the recursive "travel" on their first require (so anything they accumulate later is not protected).

cpojer commented:
Yes, I get that, but I'm worried it might get a hold of something that it's not supposed to, and then recursively protects the world.

eyalroth commented:
@cpojer I updated the branch by amending the original commit with the more robust fix (recursive, 2 levels deep).

I also added a new test environment option `disableGlobalsCleanup` to disable the globals cleanup in case of bugs with this new feature/bugfix.

cpojer commented:
That's awesome! Given that this might break existing tests, I feel like we should set the flag to default off in Jest 30, ask people to opt-in and report issues, and then possibly turn it on in the future.

This way people with memory issues can benefit from the fix right away, while people who don't care as much won't get disrupted with confusingly broken tests. What do you think?

lencioni commented:
I think that's a great idea!

cpojer commented:
Awesome! Can you update the message like this:

"[…] Jest deletes objects that were set on the global scope between test files to reduce memory leaks."

And: "In future versions of Jest, this behavior will change to "hard", which likely fails tests. You can change the behavior in your test configuration now to reduce memory usage."

?

eyalroth commented:
@cpojer I updated the message to something like this:
```
(node:73363) [JEST-01] DeprecationWarning: 'now' property was accessed on [Object] after it was soft deleted
  Jest deletes objects that were set on the global scope between test files to reduce memory leaks.
  Currently it only "soft" deletes them and emits this warning if those objects were accessed after their deletion.
  In future versions of Jest, this behavior will change to "hard", which will likely fail tests.
  You can change the behavior in your test configuration now to reduce memory usage.
```
(with amend + force push)

cpojer commented:
@eyalroth is this now ready for review/merge?

eyalroth commented:
@cpojer yes!
